File: src/js/Application.ts
import { InMemoryEntity } from "@mat3ra/code/dist/js/entity";
import {
    type Defaultable,
    defaultableEntityMixin,
} from "@mat3ra/code/dist/js/entity/mixins/DefaultableMixin";
import {
    HashedEntity,
    hashedEntityMixin,
} from "@mat3ra/code/dist/js/entity/mixins/HashedEntityMixin";
import {
    type NamedEntity,
    namedEntityMixin,
} from "@mat3ra/code/dist/js/entity/mixins/NamedEntityMixin";
import JSONSchemasInterface from "@mat3ra/esse/dist/js/esse/JSONSchemasInterface";
import type { JSONSchema } from "@mat3ra/esse/dist/js/esse/utils";
import type { ApplicationSchema } from "@mat3ra/esse/dist/js/types";

import {
    type ApplicationSchemaMixin,
    applicationSchemaMixin,
} from "./generated/ApplicationSchemaMixin";

export type DefaultApplicationConfig = Pick<
    ApplicationSchema,
    "name" | "shortName" | "version" | "summary" | "build"
>;

interface Application extends ApplicationSchemaMixin, NamedEntity, Defaultable, HashedEntity {}

class Application extends InMemoryEntity<ApplicationSchema> implements ApplicationSchema {
    static get jsonSchema(): JSONSchema {
        const schema = JSONSchemasInterface.getSchemaById("software/application");
        if (schema === undefined) {
            throw new Error('JSONSchemasInterface: missing schema id "software/application"');
        }
        return schema;
    }

    declare static createDefault: () => Application;

    getHashObject() {
        return {
            name: this.name,
            version: this.version,
            build: this.build,
        };
    }
}

namedEntityMixin(Application.prototype);
defaultableEntityMixin(Application);
applicationSchemaMixin(Application.prototype);
hashedEntityMixin(Application.prototype);

export default Application;

File: src/py/mat3ra/ade/application.py
from mat3ra.code.entity import InMemoryEntitySnakeCase
from mat3ra.code.mixins import HashedEntityMixin
from mat3ra.esse.models.software.application import ApplicationSchema
from mat3ra.utils.object import calculate_hash_from_object
from pydantic import ConfigDict, Field


class Application(ApplicationSchema, HashedEntityMixin, InMemoryEntitySnakeCase):
    """
    Application class representing a software application.

    Attributes:
        name: Application name (required)
        version: Application version
        build: Application build
        shortName: Short name of the application
        summary: Application's short description
        hasAdvancedComputeOptions: Whether advanced compute options are present
        isLicensed: Whether licensing is present
        isUsingMaterial: Whether this application is used for materials processing
        isDefault: Identifies that entity is defaultable
        schemaVersion: Entity's schema version
    """

    # `extra="ignore"` drops keys the schema doesn't declare (e.g. `buildConfig`
    # attached by standata) at construction time, so `calculate_hash` stays
    # aligned with the JS side, which performs the same schema-driven filtering
    # inside `Application.toJSON()`.
    model_config = ConfigDict(extra="ignore")

    shortName: str = ""
    summary: str = ""
    version: str = ""
    build: str = ""
    isUsingMaterial: bool = Field(default=False)

    @property
    def is_using_material(self) -> bool:
        return bool(self.isUsingMaterial)

    def get_short_name(self) -> str:
        return self.short_name if self.short_name else self.name

    def get_hash_object(self) -> dict:
        return {
            "name": self.name,
            "version": self.version,
            "build": self.build,
        }
